index/internal/indexer/indexer.go

```diff
@@ -0,0 +1,302 @@
+	ambiguous := make(map[string]collections.Set[label.Label])
+	for path, labels := range headersMapping {
+		switch len(*labels) {
+		case 1:
+			// Extract the only label in the collections.Set.
+			for l := range *labels {
+				headerToRule[path] = l
+				break
+			}
+		default:
+			ambiguous[path] = *labels
+		}
+	}
+
+	return IndexingResult{
+		HeaderToRule: headerToRule,
+		Ambiguous:    ambiguous,
+	}
+}
+
+// Writes the mapping of IndexingResult.HeaderToRule to disk in JSON format.
+// Labels are stored as renered strings
+func (result IndexingResult) WriteToFile(outputFile string) {
+	mappings := make(map[string]string, len(result.HeaderToRule))
+	for hdr, lbl := range result.HeaderToRule {
+		mappings[hdr] = lbl.String()
+	}
+
+	data, err := json.MarshalIndent(mappings, "", "  ")
+	if err != nil {
+		log.Fatalf("Failed to serialize header index to JSON: %v", err.Error())
+	}
+
+	os.MkdirAll(filepath.Dir(outputFile), 0644)
+	if err := os.WriteFile(outputFile, data, 0644); err != nil {
+		log.Fatalf("Failed to write index file: %v", err.Error())
+	}
+}
+
+// Prints to stdout detailed information about headers with resolved mappings and the ambigious header definitions.
+func (result IndexingResult) Show() {
+	log.Printf("Indexing result:")
+	log.Printf("Headers with mapping: %v", len(result.HeaderToRule))
+	for _, hdr := range slices.Sorted(maps.Keys(result.HeaderToRule)) {
+		log.Printf("%-80s: %v", hdr, result.HeaderToRule[hdr])
+	}
+
+	log.Printf("Ambigious headers: %v", len(result.Ambiguous))
+	for _, hdr := range slices.Sorted(maps.Keys(result.Ambiguous)) {
+		log.Printf("%-80s: %v", hdr, result.Ambiguous[hdr])
+	}
+}
+
+// Groups targets into disjoint groups based on the their defined headers.
+// Allows to find targets that contain at least 1 common header defined in their definition.
+func (module Module) GroupTargetsByHeaders() []collections.Set[*ModuleTarget] {
+	targets := module.Targets
+	var groups []collections.Set[*ModuleTarget]
+
+	// Build adjacency list: map each target index to its neighbors
+	adj := make(map[int][]int)
+	n := len(targets)
+	for i := range n {
+		for j := i + 1; j < n; j++ {
+			if (targets[i]).Hdrs.Intersects(targets[j].Hdrs) {
+				adj[i] = append(adj[i], j)
+				adj[j] = append(adj[j], i)
+			}
+		}
+	}
+
+	// DFS to find connected components
+	visited := make([]bool, n)
+	for i := range n {
+		if visited[i] {
+			continue
+		}
+		stack := []int{i}
+		component := make(collections.Set[*ModuleTarget])
+		visited[i] = true
+
+		for len(stack) > 0 {
+			curr := stack[len(stack)-1]
+			stack = stack[:len(stack)-1]
+			component[targets[curr]] = true
+			for _, neighbor := range adj[curr] {
+				if !visited[neighbor] {
+					stack = append(stack, neighbor)
+					visited[neighbor] = true
+				}
+			}
+		}
+		groups = append(groups, component)
+	}
+	return groups
+}
+
+// Given set of targets that define the same headers try to select ones that contain other targets as their direct or transitive dependencies
```

Comment:
This also, how will this be used? I get the sense that we're trying to resolve ambiguities when a header is in `hdrs` for multiple targets, but I wonder if we're trying to be too clever.